### PR TITLE
Add support for JSON output

### PR DESCRIPTION
**Pull request by txels**
Implements #2 

**Describe your changes**
Adds JSON support via the new CLI flag `--json`.

**Testing performed**
Added unit tests, plus tested locally against a few particularly large virtual environments. I've refactored the `print_tree` version to take some logic out of the presentation layer and in the model layer, so that it could be reused for the new JSON output format (and potential future options).

**Sample output**

```json
[
  {
    "name": "charset-normalizer",
    "version": "2.0.7",
    "up_to_date": false,
    "current": {
      "version": "2.0.7",
      "date": "2021-10-11T21:28:25.641146+00:00",
      "since": "a month ago"
    },
    "latest": {
      "version": "2.0.8",
      "date": "2021-11-24T19:46:01.258828+00:00",
      "since": "17 hours ago"
    }
  },
  {
    "name": "setuptools",
    "version": "57.4.0",
    "up_to_date": false,
    "current": {
      "version": "57.4.0",
      "date": "2021-07-19T01:20:16.478220+00:00",
      "since": "4 months ago"
    },
    "latest": {
      "version": "59.2.0",
      "date": "2021-11-19T01:14:16.274179+00:00",
      "since": "6 days ago"
    }
  },
  {
    "name": "black",
    "version": "21.11b1",
    "up_to_date": true,
    "current": {
      "version": "21.11b1",
      "date": "2021-11-18T03:58:02.282075+00:00",
      "since": "7 days ago"
    },
    "latest": {
      "version": "21.11b1",
      "date": "2021-11-18T03:58:02.282075+00:00",
      "since": "7 days ago"
    }
  },
...
]
```

### CODE REVIEW

File: src/pypi_changes/_print.py [hunk 1]
```diff
@@ -32,23 +33,55 @@ def print_tree(distributions: Iterable[Package], options: Options) -> None:
...
-        if last_release is not None:  # pragma: no branch
-            if last_release["version"] != pkg.version and pkg.info is not None:
-                for a_version, releases in pkg.info["releases"].items():  # pragma: no branch
-                    first_release_at = releases[0]["upload_time_iso_8601"]
-                    if a_version == pkg.dist.version and first_release_at is not None:
-                        text.append(" ")
-                        text.append(naturaltime(now - first_release_at), "green")
-                        break
-                text.append(f" remote {last_release['version']}", "red")
-            if last_release["upload_time_iso_8601"] is not None:  # pragma: no branch
```
Comment (txels) [inline src/pypi_changes/_print.py:45]: I'd actually like to take this refactor further and move all non-presentational logic elsewhere - a function that returns the processed dataset so that it can be fed to output formatters (rich text, json) but also consumed by 3rd party tools that use `pypi-changes` as a library.

The main point of my contribution is to improve composability of `pypi-changes` so that it can be more easily used in broader scoped tools.

File: src/pypi_changes/__init__.py [hunk 2]
```diff
@@ -15,8 +15,11 @@
 def main(args: Sequence[str] | None = None) -> int:
     options = parse_cli_arguments(args)
     distributions = collect_distributions(options)
-    info = pypi_info(distributions, options)
-    print_tree(info, options)
+    info = list(pypi_info(distributions, options))
```
Comment (gaborbernat) [inline src/pypi_changes/__init__.py:18]: I personally would prefer to keep this an iterator, and only materialize it later where it is needed to be a list.

File: .gitignore [hunk 3]
```diff
@@ -7,3 +7,4 @@ __pycache__
 /src/pypi_changes/_version.py
 /src/*.egg-info
 /.*_cache
+/.venv
```
Comment (gaborbernat) [inline .gitignore:10]: IMHO this is out of scope for this gitignore. Such configuration should be in your global git config and not in each individual project.

File: src/pypi_changes/_cli.py [hunk 4]
```diff
@@ -43,6 +43,8 @@ def _define_cli_arguments() -> ArgumentParser:
     cache_help = "seconds how long requests should be cached (pass 0 to bypass the cache, -1 to cache forever)"
     parser.add_argument("--cache-duration", "-d", default=3600, type=int, help=cache_help, metavar="SEC")
 
+    parser.add_argument("--json", action="store_true", default=False, help="Show output in JSON format")
```
Comment (gaborbernat) [inline src/pypi_changes/_cli.py:46]: Would prefer to make the presentation a mutually exclusive flag combination, where the default is the current state so it's easily extensible.

File: src/pypi_changes/_pkg.py [hunk 5]
```diff
@@ -48,6 +48,15 @@ def version(self) -> str:
     def path(self) -> Path:
         return self.dist._path  # type: ignore # it exists
 
+    @property
+    def current_release(self) -> dict[str, Any]:
+        if self.info is not None:
+            release_info = self.info["releases"].get(self.version)
+            if release_info is not None:
+                return release_info[0] or {}
+        # return empty version info if not matching
```
Comment (gaborbernat) [inline src/pypi_changes/_pkg.py:57]: Can you move this at the right side of the return below?

File: src/pypi_changes/_print.py (hunk 1 above)
Comment (gaborbernat) [reply src/pypi_changes/_print.py:45]: I'm open to this.

File: src/pypi_changes/_print.py [hunk 6]
```diff
@@ -32,23 +33,55 @@ def print_tree(distributions: Iterable[Package], options: Options) -> None:
...
+        current_release = pkg.current_release
+        current_release_at = current_release.get("upload_time_iso_8601")
+        last_release = pkg.last_release or {}
+        last_release_at = pkg.last_release_at
+
+        if current_release_at is not None:
+            text.append(" ")
+            text.append(naturaltime(now - current_release_at), "green")
+        if pkg.version != last_release.get("version"):
+            text.append(f" remote {last_release.get('version')}", "red")
```
Comment (gaborbernat) [inline src/pypi_changes/_print.py:46]: How come we needed to change the rich layer to add the JSON presentation?

File: src/pypi_changes/_print.py [hunk 7]
```diff
@@ -32,23 +33,55 @@ def print_tree(distributions: Iterable[Package], options: Options) -> None:
...
+            if last_release_at is not None:  # pragma: no branch
                 text.append(" ", "white")
-                text.append(naturaltime(now - last_release["upload_time_iso_8601"]), "green")
+                text.append(naturaltime(now - last_release_at), "green")
         tree.add(text)
     rich_print(tree)
 
 
+def release_info(release: dict[str, Any] | None, now: datetime) -> dict[str, Any]:
```
Comment (gaborbernat) [inline src/pypi_changes/_print.py:54]: I'd prefer to make the _print a package that has a two modules rich and json. And each file can only contain it's own logic. Can you do that please?

**Review by gaborbernat — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 3 above)
Comment (txels) [reply .gitignore:10]: Ah I didn't know about global git ignores, that's a great tip.

File: src/pypi_changes/_cli.py (hunk 4 above)
Comment (txels) [reply src/pypi_changes/_cli.py:46]: Makes sense. How do you want to call that flag? `--output`, `--format`? And how do you want to call the current format that uses rich for console?

File: src/pypi_changes/_print.py (hunk 6 above)
Comment (txels) [reply src/pypi_changes/_print.py:46]: The output hasn't changed to the best of my knowledge. This is a refactor to try and get all of the common "data analysis and aggregation" logic outside of the presentational layer so that it can be used by the various output formats (rich, JSON, and potential future additions).

File: src/pypi_changes/_print.py (hunk 7 above)
Comment (txels) [reply src/pypi_changes/_print.py:54]: Sure.

**Comment by txels** [comment]
Thanks for the review and the gitignore tip! I'm busy and on a business trip so I don't think I'll tackle this before next week.

**Comment by gaborbernat** [comment]
@txels do you still intend to finish this?

**Comment by txels** [comment]
> @txels do you still intend to finish this?

Hey, yeah - I was waiting for some feedback from my latest replies (e.g. https://github.com/gaborbernat/pypi_changes/pull/5#discussion_r764793216) before going forward :grimacing: as I'd prefer to do all the work in a single sitting. I'd be happy to push this through but it's up to you. I had mixed feelings re: how favorable you were to this PR, TBH.

**Comment by gaborbernat** [comment]
I'm happy to accept this change we just need to bikeshed on how it would operate. Do you still intend to action on the outstanding points and fix the merge conflicts?

File: src/pypi_changes/_print.py (hunk 7 above)
Comment (txels) [reply src/pypi_changes/_print.py]: I've done the split into `_print.json` and `_print.rich`. Not sure about what naming conventions you want to follow for these modules.

File: src/pypi_changes/_print.py (hunk 1 above)
Comment (txels) [reply src/pypi_changes/_print.py:51]: I will not do this as part of this PR to expedite its acceptance.

**Comment by txels** [comment]
I've rebased and done all the changes you suggested @gaborbernat - I've also fixed an incorrect mock that made `pytest` consistently fail locally.

I'm bumping into this issue in CI, where coverage report is failing with a generated file: 

```
packages/coverage/report.py:87: CoverageWarning: Couldn't parse '/Users/runner/work/pypi_changes/pypi_changes/src/pypi_changes/_version.py': No source for code: '/Users/runner/work/pypi_changes/pypi_changes/src/pypi_changes/_version.py'. (couldnt-parse)
coverage._warn(msg, slug="couldnt-parse")
```

...not sure what is the best way to proceed, as I don't understand why this is an issue on my branch in particular.

**Comment by txels** [comment]
Still one open pep8 issue `src/pypi_changes/_print/json.py:7:22: SC200 Possibly misspelt word: 'naturaldelta'` - how do you want to go about it?

**Comment by txels** [comment]
OK done for today, let me know what you think of the state of this so far and next steps.

File: src/pypi_changes/__init__.py (hunk 2 above)
Comment (txels) [reply src/pypi_changes/__init__.py]: Done

File: .gitignore (hunk 3 above)
Comment (txels) [reply .gitignore]: Done

File: src/pypi_changes/_cli.py (hunk 4 above)
Comment (txels) [reply src/pypi_changes/_cli.py]: Done

**Comment by txels** [comment]
@gaborbernat this is now green and conflict free. I don't think I want to spend much more time on this, it's already long gone into a zone of "diminishing returns" for me.

File: src/pypi_changes/_print/json.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
+from __future__ import annotations
+
+import json
+from datetime import datetime, timezone
+from typing import Any, Iterable
+
+from humanize import naturaldelta  # noqa: SC200
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply src/pypi_changes/_print/json.py:7]: Instead, you should add those to the whitelist.txt at the root of the project?

File: src/pypi_changes/_print/__init__.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
+from __future__ import annotations
+
+from datetime import datetime
+from typing import Iterable
+
+from .._cli import Options
+from .._pkg import Package
+
+
+class Reversor:
```
Comment (gaborbernat) [inline src/pypi_changes/_print/__init__.py:10]: This probably could/should be private so named _Reversor? or nested within the method below?

File: tests/test_print rich.py [hunk 10]
```diff
@@ -73,8 +73,8 @@ def test_print_alphabetical(capsys: CaptureFixture[str], option_simple: Options,
     assert output[0].startswith("🐍 Distributions within")
     assert output[1] == sys.executable or output[0].endswith(sys.executable)
     assert output[-4:] == [
-        "├── a 1 remote 2 a month ago",
```
Comment (gaborbernat) [inline tests/test_print rich.py:76]: Wonder why this change?

**Review by gaborbernat — CHANGES_REQUESTED** [review]
Otherwise looks good 👍

File: src/pypi_changes/_print/__init__.py (hunk 9 above)
Comment (txels) [reply src/pypi_changes/_print/__init__.py:10]: This was already in the original code.

File: tests/test_print rich.py (hunk 10 above)
Comment (txels) [reply tests/test_print rich.py:76]: The original issue was that you were using `naturaltime` whose type signature accepts `datetime|int`, which mypy was flagging. Looking into the `humanize` library, I realised that we should move to use `naturaldelta` that has the right signature (takes a `timedelta`). However, that function doesn't include `ago` in the output.

File: src/pypi_changes/_print/__init__.py (hunk 9 above)
Comment (txels) [reply src/pypi_changes/_print/__init__.py:10]: Done

File: tests/test_print rich.py (hunk 10 above)
Comment (txels) [reply tests/test_print rich.py:76]: TBH looking at the code of `humanize` the typing is inconsistent, you may consider to contribute a patch there.

https://github.com/python-humanize/humanize/blob/main/src/humanize/time.py#L208 <<< `naturaltime` signature doesn't take a timedelta - however, internally it accepts it https://github.com/python-humanize/humanize/blob/main/src/humanize/time.py#L238

**Review by gaborbernat — APPROVED** [review]

**Comment by gaborbernat** [comment]
Thanks for sticking through it! Now should be released.
